src/main/java/org/jenkinsci/remoting/util/VersionNumber.java

```diff
@@ -0,0 +1,579 @@
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.Iterator;
+import java.util.List;
+import java.util.ListIterator;
+import java.util.Locale;
+import java.util.Properties;
+import java.util.Stack;
+import java.util.regex.Matcher;
+import java.util.regex.Pattern;
+
+/**
+ * Immutable representation of a version number based on the Mercury version numbering scheme.
+ *
+ * {@link VersionNumber}s are {@link Comparable}.
+ *
+ * <h2>Special tokens</h2>
+ * <p>
+ * We allow a component to be not just a number, but also "ea", "ea1", "ea2".
+ * "ea" is treated as "ea0", and eaN &lt; M for any M &lt; 0.
+ *
+ * <p>
+ * '*' is also allowed as a component, and '*' &lt; M for any M &lt; 0.
+ *
+ * <p>
+ * 'SNAPSHOT' is also allowed as a component, and "N.SNAPSHOT" is interpreted as "N-1.*"
+ *
+ * <pre>
+ * 2.0.* &lt; 2.0.1 &lt; 2.0.1-SNAPSHOT &lt; 2.0.0.99 &lt; 2.0.0 &lt; 2.0.ea &lt; 2.0
+ * </pre>
+ *
+ * This class is re-implemented in 1.415. The class was originally introduced in 1.139
+ *
+ * @since 1.139
+ * @author Stephen Connolly (stephenc@apache.org)
+ * @author Kenney Westerhof (kenney@apache.org)
+ * @author Hervé Boutemy (hboutemy@apache.org)
+ */
+public class VersionNumber implements Comparable<VersionNumber> {
+
+    private static final Pattern SNAPSHOT = Pattern.compile("^.*((?:-\\d{8}\\.\\d{6}-\\d+)|-SNAPSHOT)( \\(.*\\))?$");
+
+    private String value;
+
+    private String snapshot;
+
+    private String canonical;
+
+    private ListItem items;
+
+    private interface Item {
+        int INTEGER_ITEM = 0;
+
+        int STRING_ITEM = 1;
+
+        int LIST_ITEM = 2;
+
+        int WILDCARD_ITEM = 3;
+
+        // Cannot set name to compareTo because FindBugs will think this interface extend Comparable and fire CO_ABSTRACT_SELF bug
+        int compare(Item item);
+
+        int getType();
+
+        boolean isNull();
+    }
+
+    /**
+     * Represents a wild-card item in the version item list.
+     */
+    private static class WildCardItem implements Item {
+
+        public int compare(Item item) {
+            if (item==null) // 1.* ( > 1.99) > 1
+                return 1;
+            switch (item.getType()) {
+                case INTEGER_ITEM:
+                case LIST_ITEM:
+                case STRING_ITEM:
+                    return 1;
+                case WILDCARD_ITEM:
+                    return 0;
+                default:
+                    return 1;
+            }
+        }
+
+        public int getType() {
+            return WILDCARD_ITEM;
+        }
+
+        public boolean isNull() {
+            return false;
+        }
+
+        @Override
+        public String toString() {
+            return "*";
+        }
+    }
+
+    /**
+     * Represents a numeric item in the version item list.
+     */
+    private static class IntegerItem
+            implements Item {
+        private static final BigInteger BigInteger_ZERO = new BigInteger("0");
```

Comment:
No need for this constant. You can refer to [`BigInteger.ZERO`](https://docs.oracle.com/javase/8/docs/api/java/math/BigInteger.html#ZERO).